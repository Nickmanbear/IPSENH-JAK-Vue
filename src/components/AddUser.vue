<template>
  <div id="add-user">
    <div id="toggle">
      <p @click="toggleIsUsers = !toggleIsUsers" v-bind:class="{ active:toggleIsUsers }">Users</p>
      <p @click="toggleIsUsers = !toggleIsUsers" v-bind:class="{ active:!toggleIsUsers }">Team</p>
    </div>

    <div v-if="toggleIsUsers">
      <input v-model="selectedUsername" type="text" placeholder="Username" list="user-list">
      <datalist id="user-list" v-if="selectedUsername">
        <option v-for="user in users" :key="user.id" :value="user.username"/>
      </datalist>
      <button @click="addUser">Add to board</button>

      <p v-if="boardUsers">Current users:</p>
      <p class="deletable" v-for="user in boardUsers" :key="user.id"
         @click="() => deleteUser(user.id)">{{ user.username }}</p>
    </div>

    <div v-else>
      <input v-model="selectedTeamName" type="text" placeholder="Team name" list="team-list">
      <datalist id="team-list" v-if="selectedTeamName">
        <option v-for="team in teams" :key="team.id" :value="team.name"/>
      </datalist>
      <button @click="addTeam">Add to board</button>

      <p v-if="boardTeam">Current team:</p>
      <p class="deletable" v-if="boardTeam" @click="deleteTeam">{{ boardTeam.name }}</p>
    </div>
  </div>
</template>

<script>
import axios from '@/axiosInstance';

export default {
  name: 'AddUser',
  props: {
    boardUsers: {},
    boardTeam: {},
  },
  data() {
    return {
      toggleIsUsers: true,
      users: [],
      teams: [],
      selectedUsername: '',
      selectedTeamName: '',
    };
  },
  mounted() {
    this.getUsers();
    this.getTeams();
  },
  methods: {
    getUsers() {
      axios.get('/user')
        .then((response) => {
          this.users = Object.keys(response.data)
            .map((key) => {
              if (!this.userOnBoard(response, key)) {
                return { id: key, username: response.data[key] };
              }
              return null;
            });
          this.users = this.users.filter((user) => (user !== null));
        });
    },
    userOnBoard(response, key) {
      return this.boardUsers.find(
        (user) => user.username === response.data[key],
      );
    },
    getTeams() {
      axios.get('/team')
        .then((response) => {
          this.teams = response.data;
        });
    },
    addUser() {
      const selectedUser = this.users.find(
        (user) => user.username === this.selectedUsername,
      );
      if (!selectedUser) {
        alert('This user doesn\'t exist.');
        return;
      }
      const userId = selectedUser.id;
      const boardId = parseInt(this.$route.params.id, 10);

      axios.post(`board/user/${boardId}/${userId}`)
        .then(() => {
          this.$emit('refresh');
          this.selectedUsername = '';
        });
    },
    addTeam() {
      const selectedTeam = this.teams.find(
        (team) => team.name === this.selectedTeamName,
      );
      if (!selectedTeam) {
        alert('This team doesn\'t exist.');
        return;
      }
      const teamId = selectedTeam.id;
      const boardId = parseInt(this.$route.params.id, 10);

      axios.post(`board/team/${boardId}/${teamId}`)
        .then(() => {
          this.$emit('refresh');
          this.selectedTeamName = '';
        });
    },
    deleteUser(userId) {
      const boardId = parseInt(this.$route.params.id, 10);
      axios.delete(`/board/user/${boardId}/${userId}`)
        .then(() => {
          this.$emit('refresh');
        });
    },
    deleteTeam() {
      const boardId = parseInt(this.$route.params.id, 10);
      axios.delete(`/board/team/${boardId}`)
        .then(() => {
          this.$emit('refresh');
        });
    },
  },
};
</script>

<style lang="scss">
#add-user {
  position: fixed;
  top: 85px;
  right: 15px;
  z-index: 90;
  border-radius: 4px;
  padding: 5px 10px;
  background-color: white;
  border: 1px solid white;
  box-shadow: 0 1px 3px rgba(0,0,0,0.12), 0 1px 2px rgba(0,0,0,0.24);

  .deletable:hover {
    text-decoration: line-through;
    cursor: pointer;
  }

  #toggle {
    text-align: center;
    padding: 5px;
    margin: 5px 0;
    background-color: #eee;
    border: 1px solid #eee;
    border-radius: 4px;

    p {
      display: inline;
      margin: 35px;
      color: #ddd;
      cursor: pointer;

      &:hover {
        color: #2c3e50;
      }
    }

    input, button {
      border-radius: 4px;
      background-color: #eee;
    }

    .active {
      color: #2c3e50;
    }
  }
}
</style>
