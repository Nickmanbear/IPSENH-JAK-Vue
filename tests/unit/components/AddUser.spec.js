import { shallowMount } from '@vue/test-utils';
import AddUser from '@/components/AddUser.vue';

const $route = { params: { id: 1 } };

jest.mock('axios', () => ({
  create: () => ({
    post: () => Promise.resolve(null),
    get: (url) => {
      let data;
      switch (url) {
        case '/team': data = [
          {
            id: 1,
            name: 'test team',
            leader: {
              id: 1,
              username: 'teamleader',
            },
            members: [
              {
                id: 1,
                username: 'teamleader',
              },
              {
                id: 2,
                username: 'member1',
              },
              {
                id: 3,
                username: 'member2',
              },
            ],
          },
          {
            id: 2,
            name: 'test team 2',
            leader: {
              id: 4,
              username: 'teamleader2',
            },
            members: [
              {
                id: 4,
                username: 'teamleader2',
              },
              {
                id: 2,
                username: 'member1',
              },
              {
                id: 3,
                username: 'member2',
              },
            ],
          },
        ];
          break;
        case '/user':
          data = {
            1: 'user1',
            2: 'user2',
            3: 'user3',
          };
          break;
        default: data = {};
          break;
      }
      return Promise.resolve({
        data,
      });
    },
    delete: () => Promise.resolve(null),
  }),
}));

describe('AddUser', () => {
  let cmp;

  beforeEach(() => {
    cmp = shallowMount(AddUser, {
      mocks: {
        $route,
      },
      propsData: {
        boardUsers: [{ id: '2', username: 'user2' }],
        boardTeam: {},
      },
      data() {
        return {
          toggleIsUsers: true,
          users: [],
          teams: [],
          selectedUsername: '',
          selectedTeamName: '',
        };
      },
    });
    window.alert = jest.fn(() => true);
  });

  it('should update selectedUsername', () => {
    cmp.find('input').setValue('test username');
    expect(cmp.vm.selectedUsername).toEqual('test username');
  });

  it('should get users', async () => {
    await cmp.vm.getUsers();
    expect(cmp.vm.users).toEqual([
      { id: '1', username: 'user1' },
      { id: '3', username: 'user3' },
    ]);
  });

  it('should get teams', async () => {
    await cmp.vm.getTeams();
    expect(cmp.vm.teams).toEqual([
      {
        id: 1,
        name: 'test team',
        leader: {
          id: 1,
          username: 'teamleader',
        },
        members: [
          {
            id: 1,
            username: 'teamleader',
          },
          {
            id: 2,
            username: 'member1',
          },
          {
            id: 3,
            username: 'member2',
          },
        ],
      },
      {
        id: 2,
        name: 'test team 2',
        leader: {
          id: 4,
          username: 'teamleader2',
        },
        members: [
          {
            id: 4,
            username: 'teamleader2',
          },
          {
            id: 2,
            username: 'member1',
          },
          {
            id: 3,
            username: 'member2',
          },
        ],
      },
    ]);
  });

  it('should add user', async () => {
    cmp.vm.users = [
      { id: '1', username: 'user1' },
      { id: '2', username: 'user2' },
      { id: '3', username: 'user3' },
    ];
    cmp.vm.selectedUsername = 'user2';

    await cmp.vm.addUser();
    expect(cmp.vm.boardUsers).toEqual([{ id: '2', username: 'user2' }]);
  });

  it('should not find add user', async () => {
    cmp.vm.users = [];
    cmp.vm.selectedUsername = 'user2';

    await cmp.vm.addUser();
    expect(cmp.vm.boardUsers).toEqual([{ id: '2', username: 'user2' }]);
  });

  it('should not find add team', async () => {
    cmp.vm.teams = [];
    cmp.vm.selectedTeamName = 'test team';

    await cmp.vm.addTeam();
    expect(cmp.vm.boardTeam).toEqual({});
  });

  it('should add team', async () => {
    cmp.vm.selectedTeamName = 'test team';

    await cmp.vm.addTeam();
    expect(cmp.vm.boardTeam).toEqual({});
  });

  it('should delete user', async () => {
    await cmp.vm.deleteUser(1);
    expect(cmp.emitted().refresh).toBeTruthy();
  });

  it('should delete team', async () => {
    await cmp.vm.deleteTeam();
    expect(cmp.emitted().refresh).toBeTruthy();
  });

  it('has the expected html structure', () => {
    expect(cmp.element).toMatchSnapshot();
  });
});
