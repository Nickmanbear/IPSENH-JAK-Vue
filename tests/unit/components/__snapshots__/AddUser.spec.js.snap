// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddUser has the expected html structure 1`] = `
<div
  id="add-user"
>
  <div
    id="toggle"
  >
    <p
      class="active"
    >
      Users
    </p>
     
    <p
      class=""
    >
      Team
    </p>
  </div>
   
  <div>
    <input
      list="user-list"
      placeholder="Username"
      type="text"
    />
     
    <!---->
     
    <button>
      Add to board
    </button>
     
    <p>
      Current users:
    </p>
     
    <p
      id="users"
    >
      user2
    </p>
  </div>
</div>
`;
